cmd/snap-confine/mount-support.c

```diff
@@ -357,10 +358,16 @@ static void sc_bootstrap_mount_namespace(const struct sc_mount_config *config)
 		// assumes to find snap-exec
 		sc_must_snprintf(dst, sizeof dst, "%s/usr/lib/snapd",
 				 scratch_dir);
-		// FIXME: take re-exec into account, i.e. this may not
-		//        actually be the path we need, we need snapd
-		//        to tell us what path to use instead.
-		const char *src = LIBEXECDIR;
+
+		// bind mount the current $ROOT/usr/lib/snapd path,
+		// where $ROOT is either "/" or the "/snap/core/current"
+		// that we are re-execing from
+		char *src = LIBEXECDIR;
+		char self[PATH_MAX + 1] = { 0, };
+		if (readlink("/proc/self/exe", self, sizeof self) > 0) {
+			src = dirname(self);
```

Comment:
Since we are guarding against kernel bugs, and we know that dirname will always return a NULL-terminated string, we can do a cheap check on the first character (we are doing this in PR 3621's bootstrap.c, for example. We should probably check for other occurences in the snap-confine sources):
```
// dirname(path) might return '.' depending on path. /proc/self/exe should always point
// to an absolute path, but let's guarantee that.
if (src[0] != '/') {
    die(...)
}